lib/SyntaxParse/SyntaxTreeCreator.cpp

```diff
@@ -164,3 +164,93 @@ SyntaxTreeCreator::lookupNode(size_t lexerOffset, syntax::SyntaxKind kind) {
   size_t length = raw->getTextLength();
   return {length, static_cast<OpaqueSyntaxNode>(raw)};
 }
+
+OpaqueSyntaxNode SyntaxTreeCreator::makeDeferredToken(tok tokenKind,
+                                                      StringRef leadingTrivia,
+                                                      StringRef trailingTrivia,
+                                                      CharSourceRange range,
+                                                      bool isMissing) {
+  // Instead of creating dedicated deferred nodes that will be recorded only if
+  // needed, the SyntaxTreeCreator always records all nodes and forms RawSyntax
+  // nodes for them. This eliminates a bunch of copies that would otherwise
+  // be required to record the deferred nodes.
+  // Should a deferred node not be recorded, its data stays alive in the
+  // SyntaxArena. This causes a small memory leak but since most nodes are
+  // being recorded, it is acceptable.
+  if (isMissing) {
+    auto Node = recordMissingToken(tokenKind, range.getStart());
+    return Node;
+  } else {
+    auto Node = recordToken(tokenKind, leadingTrivia, trailingTrivia, range);
+    return Node;
+  }
+}
+
+OpaqueSyntaxNode SyntaxTreeCreator::makeDeferredLayout(
+    syntax::SyntaxKind k, bool IsMissing,
+    const ArrayRef<RecordedOrDeferredNode> &children) {
+  auto childrenStore = ScratchAlloc.Allocate<OpaqueSyntaxNode>(children.size());
+  MutableArrayRef<OpaqueSyntaxNode> opaqueChildren =
+      llvm::makeMutableArrayRef(childrenStore, children.size());
```

Comment:
I don't think we need to allocate a new memory region for each deferred layout creation.

Reply:
It’s one of these places where I didn't put too much energy in and just kept the memory allocation scheme that’s currently used because I’ve got a further optimization up my sleeve. In this case, I’ll reuse the memory of `children` to store the `OpaqueNode`s because `OpaqueNode`s are always guaranteed to be smaller than `RecordedOrDeferredNode`s.